### PR TITLE
Add DeSelect func to SelectMenu & affected components

### PR DESCRIPTION
**Pull request by JMFURY**
Follow up from issue #196 - 

Thanks for checking out my idea! Super stoked if I can help in any way here.

### CODE REVIEW

**Comment by jeroenransijn** [comment]
This is awesome! Sorry for my late reply — have been a lot of moving parts lately on Evergreen. I don't have a ton of time to prioritize this PR atm, the code looks good. Could you provide a gif of the usage in the storybook and docs site? Let me know if you need any help with that.

**Comment by JMFURY** [comment]
Yes absolutely. Can whip something like that up soon. Will do!

**Comment by JMFURY** [comment]
@jeroenransijn  Here's a GIF of what I have so far.. 

select-multiplie-deselect

Not sure I am quite clear on whether to craft my own 'example' file under the docs tree or not, it looks like you run a command to do this? 

If that's the process should I complete the example manually in a file there (once generated/created) to match what I have here within index.stories.js ?

Thanks in advance for clarity!

**Comment by JMFURY** [comment]
Also sorry the quality sort of stinks, if you have a GIF-making toolset you use, would love to know what it is! I converted an mp4 to GIF..

**Comment by fforres** [comment]
Hijacking the thread for Gif recording info
@JMFURY for mac, [giffox](https://gifox.io/) or [kap](https://getkap.co/) + an afterpass with [giflossy](https://github.com/kornelski/giflossy) if gif ends up being too big :p

**Comment by jeroenransijn** [comment]
@JMFURY docs can be somewhat confusing. We can always add that later, the storybook is the most important thing! Everything looks good to me, the only thing we should change for moving this in is renaming `DeSelect` => `Deselect`. I found that deselect is a single word after some googling.

**Comment by JMFURY** [comment]
Awesome. Will update everywhere and commit.

**Comment by JMFURY** [comment]
@jeroenransijn  Just committed what I have, of course the example is quite a bit of code to be sitting on a component like that, without Redux or a more contextual local state to work with, it's what I could think of as a good representation. 

If you look it over and think I can consolidate for purposes of example or just in general a better way to handle it let me know. 

A lot of the time I spent really just trying to get the names formatted properly so the Button wasn't huge and rendering oddly in some of the same ways referenced here: #191

File: src/select-menu/docs/examples/SelectMenu-multi.example [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            } else {
+              selected = arrify(item.value)
+            }
+            setState({
+              selected,
+              selectedNames: selected.toString().substr(0, 25) + '...'
+            })
+          }}
+          onDeselect={item => {
+            const DeselectedItemIndex = state.selected.indexOf(item.value)
```
Comment (jeroenransijn) [inline src/select-menu/docs/examples/SelectMenu-multi.example]: `DeselectedItemIndex` => `deselectedItemIndex`. UpperCamelCase is only for class names (components) and enum like objects.

File: src/select-menu/docs/examples/SelectMenu-multi.example [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            const selectedItems = state.selected.filter(
+              (_item, i) => i !== DeselectedItemIndex
+            )
+
+            const selectedItemsLength = selectedItems.length
+            let selectedNames = ''
+            switch (true) {
```
Comment (jeroenransijn) [inline src/select-menu/docs/examples/SelectMenu-multi.example]: Wow, I just learned something new in JS. Never thought you could do this with a switch case. We should probably dumb it down to use if statements instead, I think this might not familiar for a lot of people (including me haha).

File: src/select-menu/docs/examples/SelectMenu-multi.example [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+              case selectedItemsLength === 0:
+                selectedNames = ''
+                break
+              case selectedItemsLength === 1:
+                selectedNames = selectedItems.toString()
+                break
+              case selectedItemsLength > 1:
+                selectedNames = selectedItems.toString().substr(0, 25) + '...'
```
Comment (jeroenransijn) [inline src/select-menu/docs/examples/SelectMenu-multi.example]: This is cool and a nice solution to this problem! Alternatively, I could see simply `2-n items` as the label.

File: src/select-menu/docs/examples/SelectMenu-multi.example [hunk 4]
```diff
@@ -0,0 +1,46 @@
+<Manager>
```
Comment (jeroenransijn) [inline src/select-menu/docs/examples/SelectMenu-multi.example]: This is not a blocker to merge this in, but long term we hope to migrate Evergreen over to always use `@reactions/component` in examples: https://github.com/segmentio/evergreen/issues/175. Just so you are aware next time 👍

File: src/select-menu/src/OptionsList.js [hunk 5]
```diff
@@ -33,6 +33,7 @@ export default class OptionsList extends PureComponent {
      */
     selected: PropTypes.arrayOf(PropTypes.string),
     onSelect: PropTypes.func,
+    onDeselect: PropTypes.func,
```
Comment (jeroenransijn) [inline src/select-menu/src/OptionsList.js]: Awesome, you are my hero! Thanks for renaming 👍

File: src/select-menu/stories/index.stories.js [hunk 6]
```diff
@@ -24,5 +25,51 @@ storiesOf('select-menu', module).add('SelectMenu', () => (
...
+          }}
+          onDeselect={item => {
+            const DeselectedItemIndex = state.selected.indexOf(item.value)
+            const selectedItems = state.selected.filter(
+              (_item, i) => i !== DeselectedItemIndex
+            )
+
+            const selectedItemsLength = selectedItems.length
+            let selectedNames = ''
+            switch (true) {
```
Comment (jeroenransijn) [inline src/select-menu/stories/index.stories.js]: See other comment on this statement.

File: src/select-menu/docs/examples/SelectMenu-multi.example (hunk 2 above)
Comment (JMFURY) [reply src/select-menu/docs/examples/SelectMenu-multi.example]: Ahh makes sense! I started with C++ via Arduino a couple years back actually, so this is partly that shining through. Good deal.

File: src/select-menu/docs/examples/SelectMenu-multi.example (hunk 3 above)
Comment (JMFURY) [reply src/select-menu/docs/examples/SelectMenu-multi.example]: Exactly the wisdom I was looking for!

File: src/select-menu/docs/examples/SelectMenu-multi.example (hunk 4 above)
Comment (JMFURY) [reply src/select-menu/docs/examples/SelectMenu-multi.example]: Got it! Definitely want to get this down. Will investigate.

**Comment by JMFURY** [comment]
@jeroenransijn  

All committed now, had a weird version issue with my local and VS Code.. 
Clearly I will need to get this whole process down. 

Thanks a ton for your help, insight, and patience in reviewing the code!

**Comment by jeroenransijn** [comment]
Thanks @JMFURY for putting in the first meaningful PR outside of Segment! Your help is hugely appreciated 👍

**Comment by jeroenransijn** [comment]
` evergreen-ui 3.2.0 published 🎉` Thanks @JMFURY !

**Comment by JMFURY** [comment]
@jeroenransijn So awesome! Thanks again! 

I am all-in on Evergreen. 

Cheers and congrats on the release!
